File: src/main/java/com/mrpowergamerbr/loritta/commands/vanilla/misc/WikipediaCommand.java
package com.mrpowergamerbr.loritta.commands.vanilla.misc;

import com.github.kevinsawicki.http.HttpRequest;
import com.google.common.collect.ImmutableMap;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.stream.JsonReader;
import com.mrpowergamerbr.loritta.commands.CommandBase;
import com.mrpowergamerbr.loritta.commands.CommandCategory;
import com.mrpowergamerbr.loritta.commands.CommandContext;
import net.dv8tion.jda.core.EmbedBuilder;
import org.apache.commons.lang3.StringUtils;

import java.awt.*;
import java.io.StringReader;
import java.net.URLEncoder;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

public class WikipediaCommand extends CommandBase {
    @Override
    public String getLabel() {
        return "wikipedia";
    }

    @Override
    public String getDescription() {
        return "Mostra uma versão resumida de uma página do Wikipedia";
    }

    public String getUsage() {
        return "[linguagem] conteúdo";
    }

    public List<String> getExample() {
        return Arrays.asList("Minecraft", "[en] Shantae");
    }

    public Map<String, String> getDetailedUsage() {
        return ImmutableMap.<String, String>builder()
                .put("linguagem", "*(Opcional)* Código de linguagem para procurar no Wikipédia, entre [], por padrão ele irá procurar na Wikipedia de Portugal [pt]")
                .put("conteúdo", "O que você deseja procurar no Wikipédia")
                .build();
    }

    @Override
    public CommandCategory getCategory() {
        return CommandCategory.MISC;
    }

    @Override
    public void run(CommandContext context) {
        if (context.getArgs().length >= 1) {
            String languageId = "pt";
            String inputLanguageId = context.getArgs()[0];
            boolean hasValidLanguageId = false;
            if (inputLanguageId.startsWith("[") && inputLanguageId.endsWith("]")) {
                languageId = inputLanguageId.substring(1, inputLanguageId.length() - 1);
                hasValidLanguageId = true;
            }
            try {
                String query = StringUtils.join(context.getArgs(), " ", hasValidLanguageId ? 1 : 0, context.getArgs().length);
                String wikipediaResponse = HttpRequest.get("https://" + languageId + ".wikipedia.org/w/api.php?format=json&action=query&prop=extracts&redirects=1&exintro=&explaintext=&titles=" + URLEncoder.encode(query, "UTF-8")).body();

                // Resolvi usar JsonParser em vez de criar um objeto para o Gson desparsear... a response do Wikipedia é meio "estranha"
                StringReader reader = new StringReader(wikipediaResponse);
                JsonReader jsonReader = new JsonReader(reader);
                jsonReader.setLenient(true);
                JsonObject wikipedia = new JsonParser().parse(jsonReader).getAsJsonObject(); // Base
                JsonObject wikiQuery = wikipedia.getAsJsonObject("query"); // Query
                JsonObject wikiPages = wikiQuery.getAsJsonObject("pages"); // Páginas
                Entry<String, JsonElement> entryWikiContent = wikiPages.entrySet().iterator().next(); // Conteúdo

                if (entryWikiContent.getKey().equals("-1")) { // -1 = Nenhuma página encontrada
                    context.sendMessage(context.getAsMention(true) + "Não consegui encontrar nada relacionado á **" + query + "** 😞");
                } else {
                    // Se não é -1, então é algo que existe! Yay!
                    String pageTitle = entryWikiContent.getValue().getAsJsonObject().get("title").getAsString();
                    String pageExtract = entryWikiContent.getValue().getAsJsonObject().get("extract").getAsString();

                    EmbedBuilder embed = new EmbedBuilder()
                            .setTitle(pageTitle, null)
                            .setColor(Color.BLUE)
                            .setDescription(pageExtract.length() > 512 ? pageExtract.substring(0, 509) + "..." : pageExtract);

                    context.sendMessage(embed.build()); // Envie a mensagem!
                }

            } catch (Exception e) {
                e.printStackTrace();
                context.sendMessage(context.getAsMention(true) + "**Deu ruim!**");
            }
        } else {
            context.explain();
        }
    }
}

File: loritta-discord/src/main/java/com/mrpowergamerbr/loritta/commands/vanilla/utils/WikipediaCommand.kt
package com.mrpowergamerbr.loritta.commands.vanilla.utils

import com.github.kevinsawicki.http.HttpRequest
import com.mrpowergamerbr.loritta.commands.AbstractCommand
import com.mrpowergamerbr.loritta.commands.CommandContext
import com.mrpowergamerbr.loritta.utils.Constants
import com.mrpowergamerbr.loritta.utils.jsonParser
import com.mrpowergamerbr.loritta.utils.locale.LegacyBaseLocale
import net.dv8tion.jda.api.EmbedBuilder
import net.perfectdreams.loritta.api.commands.CommandCategory
import org.apache.commons.lang3.StringUtils
import java.awt.Color
import java.net.URLEncoder
import java.util.*

class WikipediaCommand : AbstractCommand("wikipedia", category = CommandCategory.UTILS) {
	override fun getDescription(locale: LegacyBaseLocale): String {
		return locale["WIKIPEDIA_DESCRIPTION"]
	}

	override fun getUsage(): String {
		return "[linguagem] conteúdo"
	}

	override fun getExamples(): List<String> {
		return Arrays.asList("Minecraft", "[en] Shantae")
	}

	override fun getDetailedUsage(): Map<String, String> {
		return mapOf("linguagem" to "*(Opcional)* Código de linguagem para procurar no Wikipédia, entre [], por padrão ele irá procurar na Wikipedia de Portugal [pt]",
				"conteúdo" to "O que você deseja procurar no Wikipédia")
	}

	override suspend fun run(context: CommandContext,locale: LegacyBaseLocale) {
		if (context.args.isNotEmpty()) {
			var languageId = when (context.config.localeId) {
				"default" -> "pt"
				"pt-pt" -> "pt"
				"pt-funk" -> "pt"
				else -> "en"
			}

			val inputLanguageId = context.args[0]
			var hasValidLanguageId = false
			if (inputLanguageId.startsWith("[") && inputLanguageId.endsWith("]")) {
				languageId = inputLanguageId.substring(1, inputLanguageId.length - 1)
				hasValidLanguageId = true
			}
			try {
				val query = StringUtils.join(context.args, " ", if (hasValidLanguageId) 1 else 0, context.args.size)
				val wikipediaResponse = HttpRequest.get("https://" + languageId + ".wikipedia.org/w/api.php?format=json&action=query&prop=extracts&redirects=1&exintro=&explaintext=&titles=" + URLEncoder.encode(query, "UTF-8")).body()
				val wikipedia = jsonParser.parse(wikipediaResponse).asJsonObject // Base
				val wikiQuery = wikipedia.getAsJsonObject("query") // Query
				val wikiPages = wikiQuery.getAsJsonObject("pages") // Páginas
				val entryWikiContent = wikiPages.entrySet().iterator().next() // Conteúdo

				if (entryWikiContent.key == "-1") { // -1 = Nenhuma página encontrada
					context.sendMessage(Constants.ERROR + " **|** " + context.getAsMention(true) + locale["WIKIPEDIA_COULDNT_FIND", query])
				} else {
					// Se não é -1, então é algo que existe! Yay!
					val pageTitle = entryWikiContent.value.asJsonObject.get("title").asString
					val pageExtract = entryWikiContent.value.asJsonObject.get("extract").asString

					val embed = EmbedBuilder()
							.setTitle("<:wikipedia:400981794666840084> $pageTitle", null)
							.setColor(Color.BLACK)
							.setDescription(if (pageExtract.length > 512) pageExtract.substring(0, 509) + "..." else pageExtract)

					context.sendMessage(embed.build()) // Envie a mensagem!
				}

			} catch (e: Exception) {
				e.printStackTrace()
				context.sendMessage(context.getAsMention(true) + "**Deu ruim!**")
			}
		} else {
			context.explain()
		}
	}
}